### PR TITLE
feat(wasm): add parralellism in wasm API and add wasm for HLAPI

### PR DESCRIPTION
**Pull request by tmontaigu**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -302,6 +310,10 @@ test_nodejs_wasm_api_in_docker: build_nodejs_test_docker
 test_nodejs_wasm_api: build_node_js_api
 	cd tfhe && node --test js_on_wasm_tests
 
+.PHONY: test_web_js_api_parallel # Run tests for the nodejs on wasm API
```
Comment (tmontaigu) [inline Makefile:313]: Update comment

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

File: Makefile [hunk 2]
```diff
@@ -190,6 +197,15 @@ build_web_js_api: install_rs_build_toolchain install_wasm_pack
 		wasm-pack build --release --target=web \
 		-- --features=boolean-client-js-wasm-api,shortint-client-js-wasm-api,integer-client-js-wasm-api
 
+.PHONY: build_web_js_api_parallel # Build the js API targeting the web browser with parallelism support
+build_web_js_api_parallel: install_rs_check_toolchain install_wasm_pack
+	cd tfhe && \
+	rustup component add rust-src --toolchain $(RS_CHECK_TOOLCHAIN) && \
+	RUSTFLAGS="$(WASM_RUSTFLAGS) -C target-feature=+atomics,+bulk-memory,+mutable-globals" rustup run nightly-2023-04-12 \
```
Comment (IceTDrinker) [inline Makefile:204]: use the rs check toolchain variable

File: tfhe/js_on_wasm_tests/test-hlapi.js [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    let encrypted = FheUint128.encrypt_with_client_key(U128_MAX, clientKey);
+    let decrypted = encrypted.decrypt(clientKey);
+    assert.deepStrictEqual(decrypted, U128_MAX);
+
+    let serialized = encrypted.serialize();
+    let deserialized = FheUint128.deserialize(serialized);
+    let deserialized_decrypted = deserialized.decrypt(clientKey);
+    assert.deepStrictEqual(deserialized_decrypted, U128_MAX);
+
+    // Comressed
```
Comment (IceTDrinker) [inline tfhe/js_on_wasm_tests/test-hlapi.js:178]: typo Comressed -> Compressed

sed/search and replace

File: tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs [hunk 4]
```diff
@@ -50,6 +52,61 @@ pub fn decompress_seeded_lwe_ciphertext_list_with_existing_generator<
...
+        input_seeded_list.ciphertext_modulus(),
+        output_list.ciphertext_modulus(),
+    );
+
+    let ciphertext_modulus = output_list.ciphertext_modulus();
+
+    let bytes_per_child = (Scalar::BITS / 8) * (output_list.lwe_size().to_lwe_dimension().0);
+    let generators = generator
+        .par_try_fork(output_list.lwe_ciphertext_count().0, bytes_per_child)
+        .expect("Failed to fork generator");
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs:86]: there should be a par_fork_lwe_list_to_lwe that does that already, we should probably update the non parallel version above as well

File: tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs [hunk 5]
```diff
@@ -50,6 +52,61 @@ pub fn decompress_seeded_lwe_ciphertext_list_with_existing_generator<
...
+                    ciphertext_modulus.get_power_of_two_scaling_to_native_torus(),
+                );
+            }
+            *output_body.data = *body_in.data;
+        });
+}
+
 /// Decompress a [`SeededLweCiphertextList`], without consuming it, into a standard
 /// [`LweCiphertextList`].
 pub fn decompress_seeded_lwe_ciphertext_list<Scalar, InputCont, OutputCont, Gen>(
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs:112]: missing the par version of this function that would use the newly added function above, not sure it's useful

File: tfhe/src/core_crypto/entities/seeded_lwe_public_key.rs [hunk 6]
```diff
@@ -184,6 +184,27 @@ impl<Scalar: UnsignedInteger, C: Container<Element = Scalar>> SeededLwePublicKey
...
 
+    /// Consume the [`SeededLwePublicKey`] and decompress it into a standard
+    /// [`LwePublicKey`].
+    ///
+    /// See [`SeededLwePublicKey::from_container`] for usage.
+    pub fn par_decompress_into_lwe_public_key(self) -> LwePublicKeyOwned<Scalar>
+    where
+        Scalar: UnsignedTorus + Send + Sync,
+    {
+        let mut decompressed_list = LwePublicKey::new(
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/seeded_lwe_public_key.rs:195]: inherited typo this is most likely a decompressed_public_key

File: tfhe/src/js_on_wasm_api/high_level_api/integers.rs [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+}
+
+impl TryFrom<JsValue> for U256 {
+    type Error = JsError;
+
+    fn try_from(value: JsValue) -> Result<Self, Self::Error> {
+        let low_js = value.clone() & JsValue::bigint_from_str(U128_MAX_AS_STR);
+        let high_js = value >> JsValue::bigint_from_str("128");
+
+        // Since we masked the low value is will fit in u128
```
Comment (IceTDrinker) [inline tfhe/src/js_on_wasm_api/high_level_api/integers.rs:27]: is -> it

**Review by IceTDrinker — COMMENTED** [review]
A few minor comments, congrats on getting it to work, I never managed to get the threading going properly !

File: tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs (hunk 4 above)
Comment (tmontaigu) [reply tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs:86]: I did found a `par_fork_lwe_list_to_lwe` but it seems to only be available on `EncryptionGenerator` not `RandomGenerator` that we have here

Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs:86]: Ah right, that case may not have been predicted initially

File: tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs [hunk 8]
```diff
@@ -107,6 +107,34 @@ pub fn par_decompress_seeded_lwe_ciphertext_list_with_existing_generator<
...
+        output_list.ciphertext_modulus(),
+        input_seeded_list.ciphertext_modulus(),
+        "Mismatched CiphertextModulus \
+    between input SeededLweCiphertextList ({:?}) and output LweCiphertextList ({:?})",
+        input_seeded_list.ciphertext_modulus(),
+        output_list.ciphertext_modulus(),
+    );
+
+    let mut generator = RandomGenerator::<Gen>::new(input_seeded_list.compression_seed().seed);
+    decompress_seeded_lwe_ciphertext_list_with_existing_generator::<_, _, _, Gen>(
```
Comment (tmontaigu) [inline tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs:131]: needs to call `par_` version 😓

File: tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs (hunk 4 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs:86]: One of the current silent invariant is that mask byte gen does not need additional bytes as we don’t do rejection sampling, so a generator used sequentially does the job, the current pattern is forking the generator for rayon processing, we may want to have an intermediate generator MaskGenerator and NoiseGenerator which have the forking primitives and then wrap those in the encryption random generator as is currently the case and forward forking calls to the underlying new Mask and Noise gens. I guess a todo/backlog issue will do for now :/

Comment (tmontaigu) [reply tfhe/src/core_crypto/algorithms/seeded_lwe_ciphertext_list_decompression.rs:86]: Not sure I understood everything but ok 😅

**Review by IceTDrinker — APPROVED** [review]
Probably need to squash the commits I would think otherwise looks good

**Comment by IceTDrinker** [comment]
anything we can do for the timeouts ? seems a bit surprising it's so slow

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

**Comment by tmontaigu** [comment]
> anything we can do for the timeouts ? seems a bit surprising it's so slow

It timesout because the wasm part panics and aborts due to being out of memory because of the public keysize.
On my machine the test runs fine, on the CI, maybe the fact that we have to use --no-sandbox makes it so that keys are not freed between the two tests

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

**Pull request merged**